drivers/serial/uart_nrf_sw_lpuart.c

```diff
@@ -0,0 +1,658 @@
+	bool pending_rx;
+	int key;
+
+	__ASSERT_NO_MSG(data->rx_state == RX_OFF);
+
+	if (atomic_ptr_cas((atomic_ptr_t *)&data->rx_buf, NULL, buf) == false) {
+		return -EBUSY;
+	}
+
+	data->rx_len = len;
+	data->rx_timeout = timeout;
+	data->rx_state = RX_IDLE;
+
+	LOG_DBG("RX: Enabling");
+
+	key = irq_lock();
+	pending_rx = nrf_gpio_pin_read(data->rdy_pin.nrf_pin)
+		     && (data->rx_state == RX_IDLE);
+	irq_unlock(key);
+
+	if (pending_rx) {
+		activate_rx(data);
+	}
+
+	return 0;
+}
+
+static int api_rx_buf_rsp(struct device *dev, uint8_t *buf, size_t len)
+{
+	struct lpuart_data *data = get_dev_data(dev);
+
+	__ASSERT_NO_MSG((data->rx_state != RX_OFF) &&
+		 (data->rx_state != RX_TO_OFF));
+
+	if (data->rx_state == RX_TO_IDLE) {
+		data->rx_buf = buf;
+		data->rx_len = len;
+
+		if (data->rx_req) {
+			LOG_DBG("RX: Pending request. Activating RX");
+			activate_rx(data);
+		} else {
+			data->rx_state = RX_IDLE;
+			LOG_DBG("RX: Idle");
+		}
+
+		return 0;
+	}
+
+	return uart_rx_buf_rsp(data->uart, buf, len);
+}
+
+static int api_rx_disable(struct device *dev)
+{
+	struct lpuart_data *data = get_dev_data(dev);
+
+	data->rx_state = RX_TO_OFF;
+
+	return uart_rx_disable(data->uart);
+}
+
+static int lpuart_init(struct device *dev)
+{
+	struct lpuart_data *data = get_dev_data(dev);
+	const struct lpuart_config *cfg = get_dev_config(dev);
+	struct device *req_port;
+	struct device *rdy_port;
+	int err;
+
+	data->uart = device_get_binding(cfg->uart_name);
+	if (data->uart == NULL) {
+		return -ENODEV;
+	}
+
+	req_port = device_get_binding(cfg->req_port_name);
+	if (req_port == NULL) {
+		return -ENODEV;
+	}
+
+	rdy_port = device_get_binding(cfg->rdy_port_name);
+	if (rdy_port == NULL) {
+		return -ENODEV;
+	}
+
+	err = ctrl_pin_configure(&data->req_pin, req_port, cfg->req_pin, true);
+	if (err < 0) {
+		return -EINVAL;
+	}
+
+	err = ctrl_pin_configure(&data->rdy_pin, rdy_port, cfg->rdy_pin, false);
+	if (err < 0) {
+		return -EINVAL;
+	}
+
+	k_timer_init(&data->tx_timer, tx_timeout, NULL);
+	k_timer_user_data_set(&data->tx_timer, dev);
+
+	return uart_callback_set(data->uart, uart_callback, (void *)dev);
+}
+
+static int api_poll_in(struct device *dev, unsigned char *p_char)
+{
+	return -ENOTSUP;
+}
+
+static void api_poll_out(struct device *dev, unsigned char out_char)
+{
+	/* not supported */
+}
+
+static int api_configure(struct device *dev, const struct uart_config *cfg)
+{
+	struct lpuart_data *data = get_dev_data(dev);
+
+	if (cfg->flow_ctrl != UART_CFG_FLOW_CTRL_NONE) {
+		return -ENOTSUP;
+	}
+
+	return uart_configure(data->uart, cfg);
+}
+
+static int api_config_get(struct device *dev, struct uart_config *cfg)
+{
+	struct lpuart_data *data = get_dev_data(dev);
+
+	return uart_config_get(data->uart, cfg);
+}
+
+#define DT_DRV_COMPAT nordic_nrf_sw_lpuart
+
+static const struct lpuart_config lpuart_config = {
+	.uart_name = DT_INST_BUS_LABEL(0),
+	.req_port_name = (DT_INST_PROP(0, req_pin) >= 32) ?
+			"GPIO_1" : "GPIO_0",
```

Comment:
Instead of these hard-coded names you should use `label` properties of the corresponding DT nodes.